### PR TITLE
[AVCe] Re-Pak support in external BRC

### PR DESCRIPTION
**Pull request by ogolikov**
--------------- !!!! Do not merge !!!! ----------------------------

### CODE REVIEW

File: _studio/mfx_lib/encode_hw/h264/include/mfx_h264_encode_hw_utils.h [hunk 1]
```diff
@@ -1287,12 +1291,12 @@ namespace MfxHwH264Encode
         virtual mfxStatus Init(MfxVideoParam  & video) = 0;
         virtual mfxStatus Reset(MfxVideoParam  & video) = 0;
         virtual void Close() = 0;
-        virtual void PreEnc(const BRCFrameParams& par, std::vector<VmeData *> const & vmeData) = 0;
-        virtual mfxU8 GetQp(const BRCFrameParams& par) = 0;
-        virtual mfxU8 GetQpForRecode(const BRCFrameParams& par, mfxU8 curQP) = 0;
-        virtual mfxF32 GetFractionalQp(const BRCFrameParams& par) = 0;
-        virtual void SetQp(const BRCFrameParams& par, mfxU32 qp) = 0;
-        virtual mfxU32 Report(const BRCFrameParams& par, mfxU32 dataLength, mfxU32 userDataLength, mfxU32 maxFrameSize, mfxU32 qp) = 0;
+        virtual void PreEnc(BRCFrameParams& par, std::vector<VmeData *> const & vmeData) = 0;
```
Comment (dmitry-gurulev) [inline _studio/mfx_lib/encode_hw/h264/include/mfx_h264_encode_hw_utils.h]: Please do not remove 'const'

File: samples/sample_common/include/brc_routines.h [hunk 2]
```diff
@@ -348,6 +348,88 @@ class ExtBRC
...
+        else if (m_frameType & MFX_FRAMETYPE_P)
+            return 1;
+        else if ((m_frameType & MFX_FRAMETYPE_REF) && (level == 0 || bLDB))
+            return 1; //low delay B
+        else
+            return 2;
+    }
+
+private:
+    std::unique_ptr<MSDKMutex> mtx;
```
[the remark this answers is not in the document]
Comment (mgonchar) [reply samples/sample_common/include/brc_routines.h]: also no need to allocate mutex on heap

File: samples/sample_common/include/brc_routines.h [hunk 3]
```diff
@@ -401,6 +483,61 @@ namespace HEVCExtBRC
...
+        m_BRC.Init = Init;
+        m_BRC.Reset = Reset;
+        m_BRC.Close = Close;
+        m_BRC.GetFrameCtrl = GetFrameCtrl;
+        m_BRC.Update = Update;
+        return MFX_ERR_NONE;
+    }
+    inline mfxStatus Destroy(mfxExtBRC & m_BRC)
+    {
+        if (m_BRC.pthis != NULL)
```
Comment (mgonchar) [inline samples/sample_common/include/brc_routines.h]: deleting null pointer is safe and well defined procedure which simply does nothing

File: _studio/mfx_lib/encode_hw/h264/include/mfx_h264_encode_hw_utils.h [hunk 4]
```diff
@@ -1031,6 +1038,22 @@ namespace MfxHwH264Encode
 
             return false;
         }
+        inline void InitBRCParams()
+        {
+            memset(&m_brcParams, 0, sizeof(m_brcParams));
```
Comment (mgonchar) [inline _studio/mfx_lib/encode_hw/h264/include/mfx_h264_encode_hw_utils.h]: ```m_brcParams = {};```

File: _studio/mfx_lib/encode_hw/h264/include/mfx_h264_encode_hw_utils.h [hunk 5]
```diff
@@ -1031,6 +1038,22 @@ namespace MfxHwH264Encode
 
             return false;
         }
+        inline void InitBRCParams()
```
Comment (mgonchar) [inline _studio/mfx_lib/encode_hw/h264/include/mfx_h264_encode_hw_utils.h]: all class member functions are inline by default

File: _studio/mfx_lib/encode_hw/h264/include/mfx_h264_encode_hw_utils.h (hunk 1 above)
Comment (mgonchar) [reply _studio/mfx_lib/encode_hw/h264/include/mfx_h264_encode_hw_utils.h]: On which line does it change? Anyway, too many ```const``` qualifiers removed

File: _studio/mfx_lib/encode_hw/h264/include/mfx_h264_encode_hw_utils.h [hunk 6]
```diff
@@ -1630,32 +1633,31 @@ namespace MfxHwH264Encode
             }
             return MFX_BRC_OK;    
         }
-        mfxU8  GetQp(const BRCFrameParams& par)
+        mfxU8  GetQp(BRCFrameParams& par)
         {
-            mfxBRCFrameParam frame_par={};
-            mfxBRCFrameCtrl  frame_ctrl={};
+            mfxBRCFrameParam &frame_par=par.frameParam;
+            mfxBRCFrameCtrl  &frame_ctrl=par.frameCtrl;
```
Comment (mgonchar) [inline _studio/mfx_lib/encode_hw/h264/include/mfx_h264_encode_hw_utils.h]: @ogolikov did you remove all ```const``` qualifiers only to avoid copies for structs above?

File: samples/sample_common/include/brc_routines.h [hunk 7]
```diff
@@ -348,6 +348,88 @@ class ExtBRC
...
+class ExtBRCUserCQP
+{
+private:
+    bool bLDB = false;
+    mfxU8 qp[3] = {0,0,0}; // qp for I, P,B
+    mfxU32 maxFrameSize[3] = {0,0,0}; //maxFrameSize for I,P,B
+
+public:
+    ExtBRCUserCQP() { mtx.reset(new MSDKMutex()); }
+    virtual ~ExtBRCUserCQP() {}
```
Comment (mgonchar) [inline samples/sample_common/include/brc_routines.h]: why this destructor is virtual?

File: samples/sample_encode/src/pipeline_encode.cpp [hunk 8]
```diff
@@ -587,10 +584,27 @@ mfxStatus CEncodingPipeline::InitMfxEncParams(sInputParams *pInParams)
...
+        }
+        else
+        {
+            HEVCExtBRC::Create(m_ExtBRC);
+            m_bUserCQPBRC = false;
+        }
        m_EncExtParams.push_back((mfxExtBuffer *)&m_ExtBRC);
     }
 #endif
+    if (bCodingOption)
```
Comment (mgonchar) [inline samples/sample_encode/src/pipeline_encode.cpp]: why did you move this ```if```?

File: samples/sample_encode/src/pipeline_encode.cpp [hunk 9]
```diff
@@ -1615,7 +1630,10 @@ void CEncodingPipeline::Close()
     MSDK_SAFE_DELETE(m_pmfxVPP);
 
 #if (MFX_VERSION >= 1024)
-    HEVCExtBRC::Destroy(m_ExtBRC);
+    if (m_bUserCQPBRC)
+        HEVCExtBRCUserCQP::Destroy(m_ExtBRC);
```
Comment (mgonchar) [inline samples/sample_encode/src/pipeline_encode.cpp]: why do you need to do it explicitly? shouldn't ```destroy``` be called in ```m_ExtBRC``` destructor?
